subsys/bluetooth/mesh/vnd/dm_srv.c

```diff
@@ -0,0 +1,474 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <bluetooth/mesh/vnd/dm_srv.h>
+#include "../model_utils.h"
+#include <dm.h>
+#include "mesh/net.h"
+#include "mesh/access.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_MESH_DEBUG_MODEL)
+#define LOG_MODULE_NAME bt_mesh_dm_srv
+#include "common/log.h"
+
+struct bt_mesh_dm_srv *dm_srv;
+
+struct settings_data {
+	struct bt_mesh_dm_cfg cfg;
+} __packed;
+
+#if CONFIG_BT_SETTINGS
+static void store_timeout(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct bt_mesh_dm_srv *srv = CONTAINER_OF(
+		dwork, struct bt_mesh_dm_srv, store_timer);
+
+	struct settings_data data = {
+		.cfg = srv->cfg,
+	};
+
+	(void)bt_mesh_model_data_store(srv->model, true, NULL, &data,
+				       sizeof(data));
+}
+#endif
+
+static void store(struct bt_mesh_dm_srv *srv)
+{
+#if CONFIG_BT_SETTINGS
+	k_work_schedule(
+		&srv->store_timer,
+		K_SECONDS(CONFIG_BT_MESH_MODEL_SRV_STORE_TIMEOUT));
+#endif
+}
+
+static uint16_t result_convert_cm(float val)
+{
+	int32_t temp = val * 100;
+
+	if (temp < 0) {
+		return 0;
+	}
+
+	return temp > UINT16_MAX ? UINT16_MAX : (uint16_t)temp;
+}
+
+static void new_entry_store(struct dm_result *result)
+{
+	dm_srv->results.last_entry_idx =
+		(dm_srv->results.last_entry_idx + 1) % dm_srv->results.entry_cnt;
```

Comment:
You can move this to the location before `if (entry->err_occurred) {` and rephrase it to avoid errors around the wraparound boundary when wraparound does not occur at a value which is a multiple of (dm_srv->results.entry_cnt - 1).

```suggestion
        dm_srv->results.last_entry_idx++;
        dm_srv->results.last_entry_idx = dm_srv->results.last_entry_idx >= dm_srv->results.entry_cnt ? 0 : dm_srv->results.last_entry_idx;
```

Reply:
I don't see why this needs to be rephrased; using the modulo operator makes it obvious that you want the value to loop around the range [0, entry_cnt)

Reply:
Nevermind. I think I was getting hungry for lunch.